### PR TITLE
#23 Add rule that dummy arguments should have specified intent

### LINKED ISSUE #23: Check dummy arguments have a specified intent
A check to test whether a subroutine or functions dummy arguments contain the intent attribute. Not having intent currently trips up the pointer initialisation checking when intent(in) is not specified for pointers. This check should therefore ideally happen before the pointer init check.

### PR DESCRIPTION
**Pull request by bilal-chughtai**
Implemented rule outlined in #23, along with some unit and system testing. I note that some of the code blocks the unit test produces while checking all permutations are invalid fortran, specifying for example two or no `intent(out)` variables for a function. I think this is fine, since the compiler will catch these issues, and logic for preventing this complicates the unit test unnecessarily, and is beyond the scope of the rule. This rule won't work find subroutines or functions that don't reside within a `program` or `module`, which is a property I'm aware some legacy code can exhibit. I'm leaving this as a known shortcoming. Closes #23.

### CODE REVIEW

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
Looks good. A report on using the rule on a sample codebase, such as LFRic, would be good evidence of function.

File: source/stylist/fortran.py [hunk 1]
```diff
@@ -174,6 +176,80 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
...
+                                         'Module_Subprogram_Part',
+                                         'Module_Subprogram']))
+        scope: Fortran2003.Block
+        for scope in scope_units:
+
+            # get the type of block
+            if type(scope) == Fortran2003.Subroutine_Subprogram:
+                unit_type = 'subroutine'
+            elif type(scope) == Fortran2003.Function_Subprogram:
+                unit_type = 'function'
```
[the remark this answers is not in the document]
Comment (bilal-chughtai) [reply source/stylist/fortran.py:205]: moved.

File: source/stylist/fortran.py [hunk 2]
```diff
@@ -174,6 +176,80 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
...
+
+            # get initialisation statement and piece containing type
+            # declarations
+            for part in scope.children:
+                if type(part) == Fortran2003.Function_Stmt or type(
+                        part) == Fortran2003.Subroutine_Stmt:
+                    stmt = part
+                if type(part) == Fortran2003.Specification_Part:
+                    specs = part
+                    # that's all we need
```
[the remark this answers is not in the document]
Comment (bilal-chughtai) [reply source/stylist/fortran.py:215]: elaborated

File: source/stylist/fortran.py [hunk 3]
```diff
@@ -174,6 +176,80 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
...
+                    break
+
+            # covert the node into a python list
+            dummy_arg_list = stmt.children[2]
+            dummy_args = []
+            if dummy_arg_list is not None:
+                dummy_args = [arg.string.lower() for arg in
+                              dummy_arg_list.children]
+
+            for spec in specs.children:
```
[the remark this answers is not in the document]
Comment (bilal-chughtai) [reply source/stylist/fortran.py:225]: Things fall over. The only way I can see this happening is if you declare a subprogram and put nothing in it, as the specification part is the first thing fparser looks for.

```
subroutine foo(bar)
end subroutine blah
```
I'll add some logic and a test to deal with this appropriately.

File: source/stylist/fortran.py [hunk 4]
```diff
@@ -174,6 +176,80 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
...
+                    if attributes is not None:
+                        for attribute in spec.children[1].children:
+                            if attribute.__class__ == \
+                                    Fortran2003.Intent_Attr_Spec:
+
+                                # if so, remove argument names from dummy_args
+                                for arg in spec.children[2].children:
+                                    arg_name = arg.children[0].string
+                                    if arg_name.lower() in dummy_args:
+                                        dummy_args.remove(arg_name.lower())
```
[the remark this answers is not in the document]
Comment (bilal-chughtai) [reply source/stylist/fortran.py:239]: This does however make testing harder - the ordering of iteration over a list is unpredictable. I think I'll keep the original.

File: system-tests/fortran/missing_intent.f90 [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+contains
+
+  subroutine test_subroutine( foo, bar, foobar, baz, qux )
+
+      implicit none
+
+      integer, intent(in), pointer :: foo
+      character, intent(inout):: bar, foobar
+      real :: BAZ(5)
+      logical, intent(inout):: qux
```
[the remark this answers is not in the document]
Comment (MatthewHambley) [reply system-tests/fortran/missing_intent.f90]: Did this get forgotten?

File: system-tests/fortran/missing_intent.f90 [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      logical, pointer :: quux => null()
+
+  end subroutine test_subroutine
+
+end module test_module
```
Comment (MatthewHambley) [inline system-tests/fortran/missing_intent.f90:34]: Now that the rule supports procedure program units they should be tested here.

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
Just two minor things to finish and then it's off to the races.

**Pull request merged**